design/proposal-controller-code-generation.md

```diff
@@ -0,0 +1,129 @@
+
+## Pipeline Architecture
+In order to decouple interpretation of a Provider API description from the code generation process, it seems useful to borrow the concept of frontend/backend layering from compilers. In a traditional compiler, the frontend is responsible for parsing source code and generating an intermediate format, often an abstract syntax tree, which the backend parses in order to perform platform specific code generation. This separation of concerns allows a compiler toolchain to interpret different languages on the frontend and target different architectures on the backend.
+
+In our case, the backend would be responsible for generating Controller boilerplate independent of the Provider. This allows different frontend implementations to be written for different Providers, or possibly using different API metadata sources for the same Provider. For instance, a frontend could analyze an OpenAPI document, cloud provider specific discovery documents, or directly parse cloud provider SDKs. In between backend and frontend we need to define an intermediate format (like an AST in a compiler).
+
+## MRS
+
+Since the CRD is the lingua franca of Kubernetes API objects, it makes intuitive sense to use a json-schema representation CRDs, with a few changes. First, we may want to simplify the schema, trimming away aspects that exist to support the Kubernetes runtime environment, like schema versioning and subresources. Second, there are additional metadata we expect to infer from the Provider, or to be overlayed into the intermediate format by developers, which we want to maintain push downstream, like documentation, or metadata indicating that a field is optional, mutable, read-only, or useful for `controller-tools` annotations. Using a tree-like data structure makes for a simpler and less heuristic parser compared to scanning through a Go source tree, and should be easier to work with in tests.  Using a json-schema representation allows us to leverage json-schema validation and libraries.  Managed Resource Schema (MRS) is the working title for this format.
+
+## Open Questions
+
+### Provider API metadata sources
+More work is needed to determine which source to use for generating the MRS for each Provider API. Some possibilities:
+
+**OpenAPI**: Potentially allows us to write the simplest frontend parser; major cloud providers either provide OpenAPI documents (Alibaba, Azure) or well-known discovery formats for which mechanical translators are available (Google), or bundle OpenAPI documents in their SDK repos for runtime support or to drive code generation (AWS).
```

Comment:
OpenAPI definitely seems like the lowest friction way to get started with this effort, but I agree the frontend needs to be pluggable as we cannot foresee all of the providers and their API metadata sources we may support in the future

Reply:
I'm already seeing this break down in the case of Google, where the OpenAPI conversions are missing endpoints, so provider-specific frontends seem very likely.